crates/factor-key-value/src/spin_cli_resolver.rs

```diff
@@ -0,0 +1,47 @@
+use crate::runtime_config::RuntimeConfigResolver;
+use crate::store::{store_from_toml_fn, MakeKeyValueStore, StoreFromToml};
+use spin_key_value::StoreManager;
+use std::{collections::HashMap, sync::Arc};
+
+#[derive(Default)]
+pub struct SpinCliRuntimeConfigResolver {
+    store_types: HashMap<&'static str, StoreFromToml>,
+}
+
+impl SpinCliRuntimeConfigResolver {
+    pub fn add_store_type<T: MakeKeyValueStore>(&mut self, store_type: T) -> anyhow::Result<()> {
+        if self
+            .store_types
+            .insert(T::RUNTIME_CONFIG_TYPE, store_from_toml_fn(store_type))
+            .is_some()
+        {
+            anyhow::bail!(
+                "duplicate key value store type {:?}",
+                T::RUNTIME_CONFIG_TYPE
+            );
+        }
+        Ok(())
+    }
+}
+
+impl RuntimeConfigResolver for SpinCliRuntimeConfigResolver {
+    fn get_store(
+        &self,
+        store_kind: &str,
+        config: toml::Table,
+    ) -> anyhow::Result<Arc<dyn StoreManager>> {
+        let store_from_toml = self
+            .store_types
+            .get(store_kind)
+            .ok_or_else(|| anyhow::anyhow!("unknown store kind: {}", store_kind))?;
+        store_from_toml(config)
+    }
+
+    fn default(&self, label: &str) -> Option<Arc<dyn StoreManager>> {
+        if label == "default" {
+            self.get_store("spin", toml::value::Table::new()).ok()
+        } else {
+            None
+        }
+    }
```

Comment:
To make this not-`spin up`-specific, maybe we could make this configurable, with a new method on the factor, e.g. `kv_factor.default_store("spin", toml::value::Table::new())`. Note also that the `Table` there will actually need a `path` for the default store derived by the runtime anyway.

Reply:
The tricky thing here is that we want to abstract away what the default label is too. For Spin it happens to be "default". Furthermore, a runtime may want to establish multiple defaults, say "cache" is always the local sqlite impl.

Reply:
@lann what do you think of this 7f7ebc33a63f9393670daadfd96bb034394497f3

Reply:
I am now trying to figure out the best way to pass the default store opts when we make a store: https://github.com/kate-goldenring/spin/blob/41a5796603c80617eebed2eac7a12d8f475bfaae/crates/trigger/src/runtime_config/key_value.rs#L91-L92

Reply:
Think I got it all situated: 318e1660113d45cbcae8b202a47e174b14a938d9